week-2/learning-outcomes.md

```diff
@@ -0,0 +1,37 @@
+# Week 2 Learning Outcomes
+
+## HTTP
+
+### Asynchronicity
+
+- [ ] **We can write code that executes asynchronously**
+- [ ] We can use callbacks to access values that aren't available synchronously
+- [ ] We can use promises to access values that aren't available synchronously
```

Comment:
We need to give students a while to understand callbacks before we move onto promises IMO

Reply:
Having tried both ways, I think that our standard approach of starting with XMLHttpRequest is the better option, as it makes it clearer what is happening

Reply:
Do you mean you tried starting with `fetch` at Beamery? The middle east campuses have been doing first-class functions -> callbacks -> async -> fetch/promises all in one day for a long time now and it seems to work for them.

What concepts does XMLHttpRequest make clearer?

Reply:
Yeah, I tried starting with fetch, but felt like it abstracted away too much of what I wanted to get across. With XHR I like how the callback is called on each readystate change and how this demonstrates the lifecycle of the HTTP request. But none of that is exactly necessary...

>The middle east campuses have been doing first-class functions -> callbacks -> async -> fetch/promises all in one day for a long time now and it seems to work for them.

I think the advantage to this was to take a simpler option than using XHRs in the early stage of the course. We still need to map out how we build up understanding and concepts around async. By using promises so early, I worry that we don't give students time to understand or use callbacks.

Reply:
Yeah I understand your concern. At this point though the JS ecosystem is so saturated in promises that we lose more by not covering them early on. I don't think it's necessarily a bad thing to get higher-level concepts early on and understand more deeply once you're comfortable with getting stuff done.

Also promises still use callbacks, so the basic concept has to be understood.

What exactly were you teaching around the lifecycle of the HTTP request? We might still be able to cover how it works (or what fetch is doing under-the-hood)

Reply:
@arrested-developer It's a good point because IMO callbacks are hard to grasp at first, but important to understand (even if using promises is the norm for many tasks nowadays).

@oliverjam What about if we put a workshop in this week that shows students how to use both callbacks and promises to complete the same task? Like how the error handling workshop in week 5 gets you to write the same set of functions, but with different error handling methods?

I think it would also be worth making callbacks its own research topic, there was a really good FAC18 presentation that explained the node error first callback pattern to the group.

Reply:
I think callbacks are too important and too confusing (initially) to leave to a research topic. It's hard enough to explain even when you know exactly how to use them yourself.

I honestly don't think it will be that big an issue to have promises and callbacks side by side. The most important thing is for them to consistently get exposed to and practice using them throughout the course, so it clicks by the end.

Reply:
@arrested-developer I understand why you want to keep xhr for conceptual understanding. On the other hand, it might be a good idea to expose a cohort to promises before the testing and server weeks (possibly the the most stressful weeks of the programme). Also, going with fetch will keep the Middle East campuses on board, which I am keen to do.